packaging/requirements.py

```diff
@@ -0,0 +1,114 @@
+
+LBRACKET = L("[").suppress()
+RBRACKET = L("]").suppress()
+LPAREN = L("(").suppress()
+RPAREN = L(")").suppress()
+COMMA = L(",").suppress()
+SEMICOLON = L(";").suppress()
+AT = L("@").suppress()
+
+IDENTIFIER_C = ALPHANUM + Word("-_.")
+IDENTIFIER = ALPHANUM | (ALPHANUM + ZeroOrMore(IDENTIFIER_C) + ALPHANUM)
+
+NAME = IDENTIFIER("name")
+EXTRA = IDENTIFIER
+
+URI = Regex(r'[^ ]+')("url")
+URL = (AT + URI)
+
+EXTRAS = (LBRACKET + EXTRA + ZeroOrMore(COMMA + EXTRA) + RBRACKET)("extras")
+
+VERSION_PEP440 = Regex(Specifier._regex_str, re.VERBOSE | re.IGNORECASE)
+VERSION_LEGACY = Regex(LegacySpecifier._regex_str, re.VERBOSE | re.IGNORECASE)
+
+VERSION_ONE = VERSION_PEP440 | VERSION_LEGACY
+VERSION_MANY = VERSION_ONE + ZeroOrMore(COMMA + VERSION_ONE)
+VERSION_MANY.setParseAction(
+    lambda s, l, t: SpecifierSet(','.join(t.asList()))
+)
+VERSION_SPEC = ((LPAREN + VERSION_MANY + RPAREN) | VERSION_MANY)
+VERSION_SPEC = originalTextFor(VERSION_SPEC())("specifier")
+
+MARKER_EXPR = originalTextFor(MARKER_EXPR())("marker")
+MARKER_EXPR.setParseAction(
+    lambda s, l, t: Marker(s[t._original_start:t._original_end])
+)
+MARKER_SEPERATOR = SEMICOLON
+MARKER = MARKER_SEPERATOR + MARKER_EXPR
+
+VERSION_AND_MARKER = Optional(VERSION_SPEC) + Optional(MARKER)
+URL_AND_MARKER = URL + Optional(MARKER)
+
+NAMED_REQUIREMENT = \
+    NAME + Optional(EXTRAS) + (URL_AND_MARKER | VERSION_AND_MARKER)
+
+REQUIREMENT = stringStart + NAMED_REQUIREMENT + stringEnd
+
+
+class Requirement(object):
+
+    # TODO: Can we test whether something is contained within a requirement?
+    #       If so how do we do that? Do we need to test against the _name_ of
+    #       the thing as well as the version? What about the markers?
+    # TODO: How do we handle the "extra" marker since it's special?
+    # TODO: Can we normalize the name and extra name?
+
+    def __init__(self, requirement_string):
+        req = REQUIREMENT.parseString(requirement_string)
+
+        self._name = req.name
+        if req.url:
+            parsed_url = urlparse.urlparse(req.url)
+            if not (parsed_url.scheme and parsed_url.netloc) or (
+                    not parsed_url.scheme and not parsed_url.netloc):
+                raise InvalidRequirement("Invalid URL given")
+            self._url = req.url
+        else:
+            self._url = None
+        self._extras = req.extras.asList() if req.extras else None
+        self._specifier = req.specifier if req.specifier else None
+        self._marker = req.marker if req.marker else None
+
+    def __str__(self):
+        parts = [self._name]
+
+        if self._extras:
+            parts.append("[{0}]".format(",".join(sorted(self._extras))))
+
+        if self._specifier:
+            parts.append(str(self._specifier))
+
+        if self._url:
+            parts.append("@ {0}".format(self._url))
+
+        if self._marker:
+            parts.append("; {0}".format(self._marker))
+
+        return "".join(parts)
```

Comment:
I very much want to be able to pull out the strings of the different bits - my use case is taking a bunch of marker annotated dependencies, evaluating the marker to ge tthe ones relevant to here-and-now, and then serialise back to strings *without* the markers.

Without getting into mutable objects, something like:
 - make name, extras, specifiers, url, marker be top level properties
 - make them all either strings or things with a __str__ that looks like a (perhaps canonised) equivalent to the original

so that I can write:
"%s%s%s" % (req.name, req.extras, req.specifiers)
and have it be a legitiate dependency specifier. (I won't be passing urls to setuptools...)